src/libmain/progress-bar.cc

```diff
@@ -324,174 +603,273 @@ class ProgressBar : public Logger
+
+            state.statusLines.insert_or_assign({idBuilds, n++},
+                fmt("  %s",
+                    renderBar(builds.done, builds.failed, builds.running, builds.expected)));
+
+            for (auto & build : state.activitiesByType[actBuild].its) {
+                state.statusLines.insert_or_assign({idBuilds, n++},
+                    fmt(ANSI_BOLD "  ‣ %s (%d s)%s: %s",
+                        build.second->s,
+                        std::chrono::duration_cast<std::chrono::seconds>(now - *build.second->startTime).count(),
+                        build.second->phase ? fmt(" (%s)", *build.second->phase) : "",
+                        build.second->lastLine));
+            }
 
-        if (state.corruptedPaths) {
-            if (!res.empty()) res += ", ";
-            res += fmt(ANSI_RED "%d corrupted" ANSI_NORMAL, state.corruptedPaths);
+            state.statusLines.insert_or_assign({idBuilds, n++}, "");
         }
 
-        if (state.untrustedPaths) {
-            if (!res.empty()) res += ", ";
-            res += fmt(ANSI_RED "%d untrusted" ANSI_NORMAL, state.untrustedPaths);
+        auto verify = getActivityStats(state.activitiesByType[actVerifyPaths]);
+
+        if (verify.done || verify.expected) {
+            removeStatusLines(state, idVerifyPaths);
+
+            auto bad = state.corruptedPaths + state.untrustedPaths + verify.failed;
+
+            size_t n = 0;
+            state.statusLines.insert_or_assign(
+                {idVerifyPaths, n++},
+                fmt("%s Verify %d / %d paths",
+                    verify.done < verify.expected
+                    ? busyMark
+                    : bad
+                    ? failMark
+                    : doneMark,
+                    verify.done, verify.expected)
+                + (state.corruptedPaths ? fmt(", %d corrupted", state.corruptedPaths) : "")
+                + (state.untrustedPaths ? fmt(", %d untrusted", state.untrustedPaths) : "")
+                + (verify.failed ? fmt(", %d failed", verify.failed) : "")
+                );
+
+            state.statusLines.insert_or_assign({idVerifyPaths, n++},
+                fmt("  %s", renderBar(verify.done - bad, bad, verify.running, verify.expected)));
+
+            for (auto & build : state.activitiesByType[actVerifyPath].its) {
+                state.statusLines.insert_or_assign({idVerifyPaths, n++},
+                    fmt(ANSI_BOLD "  ‣ %s", build.second->s));
+            }
+
+            state.statusLines.insert_or_assign({idVerifyPaths, n++}, "");
         }
+    }
 
-        return res;
+    void draw(State & state, std::optional<std::string_view> msg = {})
+    {
+        state.haveUpdate = false;
+        if (!state.active) return;
+
+        auto width = getWindowSize().second;
+        if (width <= 0) width = std::numeric_limits<decltype(width)>::max();
+
+        std::string s;
+
+        for (size_t i = 1; i < state.prevStatusLines; ++i)
+            s += "\r\e[K\e[A";
```

Comment:
What are these escape sequences? Maybe drop a comment what they are doing and why they are here.

Reply:
I was curious so I looked it up. Posting the answer in case anybody else wonders.

Anything of the form `\e[A` is an ansi escape, that is interpreted by the terminal. It's the same kind that is used to output color. '\r' is a carriage return, which puts the cursor back at the beginning of the line.

All put together we have:
- `\r` return the cursor at the beginning of the line
- `\e[K` erases the line from the cursor
- `\e[A` moves the cursor one line up

So, for each line we had in the previous status line, we erase a full line and move the cursor up. Ie, we delete all we printed previously.